packages/taquito/src/contract/rpc-estimate-provider.ts

```diff
@@ -247,14 +298,97 @@ export class RPCEstimateProvider extends OperationEmitter implements EstimationP
    *
    * @param Estimate
    */
-  async registerDelegate(params: RegisterDelegateParams) {
-    const DEFAULT_PARAMS = await this.getAccountLimits(await this.signer.publicKeyHash());
+  async registerDelegate(params: RegisterDelegateParams, estimationOptions: EstimationOptions = { includeRevealOperation: true }) {
+    const pkh = await this.signer.publicKeyHash()
+    const DEFAULT_PARAMS = await this.getAccountLimits(pkh);
     const op = await createRegisterDelegateOperation(
       { ...params, ...DEFAULT_PARAMS },
-      await this.signer.publicKeyHash()
+      pkh
     );
-    return (
-      await this.createEstimate({ operation: op, source: await this.signer.publicKeyHash() })
-    )[0];
+    const isRevealNeeded = await this.isRevealOpNeeded([op], pkh);
+    const ops = isRevealNeeded ? await this.addRevealOp([op], pkh) : op;
+    const estimateProperties = await this.prepareEstimate({ operation: ops, source: pkh });
+    if (!estimationOptions.includeRevealOperation && isRevealNeeded) {
+      estimateProperties.shift();
+    }
+    return this.createEstimateObject(estimateProperties);
+  }
+
+  /**
+   *
+   * @description Estimate gasLimit, storageLimit and fees to reveal the current account
+   *
+   * @returns An estimation of gasLimit, storageLimit and fees for the operation or undefined if the account is already revealed
+   *
+   * @param Estimate
+   */
+  async reveal(params?: RevealParams) {
+    const pkh = await this.signer.publicKeyHash();
+    if (await this.isAccountRevealRequired(pkh)) {
+      const DEFAULT_PARAMS = await this.getAccountLimits(pkh);
+      const op = await createRevealOperation({
+        ...params, ...DEFAULT_PARAMS
+      },
+        pkh,
+        await this.signer.publicKey()
+      )
+      const estimateProperties = await this.prepareEstimate({ operation: op, source: pkh });
+      return this.createEstimateObject(estimateProperties);
+    }
+  }
+
+  private isRevealRequiredForOpType(op: RPCOperation[]) {
+    let opRequireReveal = false;
+    for (const operation of op) {
+      if (isOpRequireReveal(operation)) {
+        opRequireReveal = true;
+      }
+    }
+    return opRequireReveal;
+  };
+
+  private async isRevealOpNeeded(op: RPCOperation[], pkh: string) {
+    return (!await this.isAccountRevealRequired(pkh) || !this.isRevealRequiredForOpType(op)) ? false : true;
+  }
+
+  private async addRevealOp(op: RPCOperation[], pkh: string) {
+    op.unshift(
+      await createRevealOperation({
+        ...{ fee: DEFAULT_FEE.REVEAL, gasLimit: DEFAULT_GAS_LIMIT.REVEAL, storageLimit: DEFAULT_STORAGE_LIMIT.REVEAL }
+      },
+        pkh,
+        await this.signer.publicKey())
+    )
+    return op;
+  }
+
+  // Create an Estimate instance that include both reveal and operation fees
+  private createEstimateObject(estimateProperties: EstimateProperties[]) {
```

Comment:
Should we move this to the Estimate class? What if we have to duplicate this logic in the native estimate class or any other implementation?

Reply:
Done, thanks for pointing this.